### PR TITLE
Added bytecode to project

### PR DESCRIPTION
**Pull request by SirPaws**
I've finished writing the bytecode format, **this is not ready to be merged** but I wanted open a pull request so people can check if I've forgotten stuff or if I can improve upon it

as I mentioned this isn't ready to be merged, I'll be connecting the bytecode with the language tomorrow but for now, I wanted to show the progress I've made and make an early PR.
also, this next bit is more for the report but I'll write it as a 'note' here.

# design
in reality, I didn't really consider the design of the bytecode I just used what I've seen before, but with that said I was thinking of some of the instruction sets and intermediate representations I've had a look at before 
in this case, I was thinking of the RISC-V/X86 instruction sets, MSIL/CIL[^1] and MIR[^2]

I've included the textual format[^3] that I used as a guide for implementing it (**;** is the start of a comment)

oh and also I haven't had time to implement a binary output format but I've thought about it 
```
HEADER
SYMBOL_TABLE
INSTRUCTIONS ....
```
the header would just be an identifier to make sure it's actually bytecode we're reading, something like 4 bytes saying **CODE** in ascii.
the symbol table is not actual language symbols but identifiers for variables, so it's essentially just a list of null-terminated strings.
the next bit is the instructions these would be read as packets of 13 bytes `MNEMONIC | ARG0 | ARG1 | ARG2` 
- 1 byte mnemonic (instruction or whatever you want to call it)
- 3 * 32-bit unsigned integer representing an offset in the symbol table

the only caveat is that `func` and `call` need special handling, but they are still loaded in the same way
for `func` the first integer is the number of parameters and the second integer will be the number of instructions that this function uses.
the following bytes would then be the parameters (still 32-bit unsigned integers) followed by the instructions (for `func` only)
```
func num_params, num_instructions, 0
parameter_names....
instructions....
```

for call the first integer is the 'return address or the variable where the result is stored, the next is the actual function we want to call, followed by the number of arguments. 
same as with `func` the next bytes would then be the parameters
```
call return_variable, function, num_params
parameter_names...
``` 

[^1]: https://github.com/vnmakarov/mir
[^2]: https://docs.microsoft.com/en-us/dotnet/standard/managed-code
[^3]: the example_bytecode.bt file

### CODE REVIEW

**Review by miningape — APPROVED** [review]
Looks good to me, I just wished I knew you were making this then I wouldnt have made my own shitty bytecode :P

I'll use both in my report (quick and dirty vs proper implementation) so can we set up a call next week where I ask you a few questions?

**Comment by miningape** [comment]
Also I didnt see any code that explicitely tells me your handling closures, are you?

File: bytecode/src/execute.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        Mnemonics.LOGICAL_AND          : logical_and_closure.evaluate,
+        Mnemonics.LOGICAL_OR           : logical_or_closure.evaluate,
+
+        # arithmetic
+        Mnemonics.ADDITION       : addition_closure.evaluate,
+        Mnemonics.SUBTRACTION    : subtraction_closure.evaluate,
+        Mnemonics.DIVISION       : division_closure.evaluate,
+        Mnemonics.MULTIPLICATION : multiplication_closure.evaluate,
+    }
+    assert len(execute_byte.keys()) == len(Mnemonics)
```
Comment (miningape) [inline bytecode/src/execute.py:160]: what is this for?

Comment (SirPaws) [reply bytecode/src/execute.py:160]: I wanted something like a switch in python and the assertion is just to make sure it's exhaustive

**Comment by SirPaws** [comment]
okay so the bytecode stuff is now finished (still no binary format) it's fully connected with the language.
I wrote a 'changelog' in the commit description but I'll also post it here so it's easier to see 

1.  moved bytecode files from temporary folder into actual location
2.  added bytecode generator that takes in a file and generates bytecode
3.  renamed 'bytecode' Literal to Value to not confuse with AST
4.  added missing bytecodes. namely NotEquals, unary operators (!, -, +) and some dealing with scope (Frame, Raze)
5.  edited main.pych to test more functionality
6.  added printing of bytecode
7.  added extra command line arguments to allow generating/executing bytecode, also for printing said bytecode
8.  did some minor cleanup, but more could be used
9.  removed bytecode temporary folder (I kept the example_bytecode.bt file because it shows some of the initial ideas)

**Comment by SirPaws** [comment]
okay so now all the new features are supported, one small note is that I made arrays very forgiving.
right now the bytecode would see code like this as valid
```js
let a = [1, 2, 3]
a[print] = 5
```
It's honestly trivial to do it properly but I don't feel like it and if someone else want's to do it go ahead

**Comment by SirPaws** [comment]
there's also currently a bug with variables, but it is caused by pythons handling of variables and I am not knowledgable enough to be able to fix it currently, and I honestly can't be bothered

File: bytecode/src/execute.py (hunk 1 above)
Comment (miningape) [reply bytecode/src/execute.py]: Yeah I wanted to use that but there isnt a stable version for apple silicon and I didnt feel like building from source

File: src/pychart/_interpreter/visitors/bytecode_generator.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
+
+from typing import Any, Dict, List, Callable
+
+from src.pychart._interpreter.ast_nodes.statement import *
```
[the remark this answers is not in the document]
Comment (miningape) [reply src/pychart/_interpreter/visitors/bytecode_generator.py:4]: nevermind paws used some tricks that arent exactly type friendly and I dont wanna make changes so i'm not gonna be making the pr

**Comment by spralja** [comment]
You can make it a draft in the menu to the left

**Pull request merged**